### DISCUSSION TITLE
Is there a help command builder for slash commands?

### DISCUSSION THREAD

**Discussion by Andre601** [Chew/JDA-Chewtils#15] [Q&A] [answered]
I was wondering if there was a builder or similar for creating a help command with Slash commands.
By that I don't necessarily mean a setup that automatically adds a `/help` command (That would be stupid).

But more a method or tool that would create and return a list of the registered commands, with their arguments provides.

Example:
Assume I have `/img cat [amount]` and `/img dog [amount]` that return cat and dog images respectively (`[amount]` would be optional.

Now calling a specific method in chewtils would create and return this list.

```
/img cat [amount]
/img dog [amount]
```

And while I'm at it, does Chewtil offer categorization of the commands?
Like that I can put some commands under the category fun, others under settings, etc. and with the above system would it list them by category.

**Comment by Chew** [comment, marked as answer]
The easiest way I can think of is just reverse-engineering the existing help builder.

```java
// Inside an execute(SlashCommandEvent event)
var client = event.getClient();
var commands = client.getSlashCommands();

// Get categories
StringBuilder builder = new StringBuilder("**" + event.getSelfUser().getName() + "** commands:\n");
Command.Category category = null;
for (Command command : event.getClient().getCommands()) {
    if (command.isHidden() || command.isOwnerCommand()) continue;
    if (!Objects.equals(category, command.getCategory())) {
        category = command.getCategory();
        builder.append("\n\n  __").append(category == null ? "No Category" : category.getName()).append("__:\n");
    }
    builder.append("\n`").append(event.getPrefix()).append(command.getName())
             .append(command.getArguments() == null ? "`" : " " + command.getArguments() + "`")
             .append(" - ").append(command.getHelp());
}
event.reply(builder.toString()).queue();
```
Granted, you would need to fill out `this.args` as you would do for the normal `!help` response, but merely for documentation.

**Reply by Andre601** [reply]
`event.getClient()` is not a thing :^)

You probably meant `this.getClient()` instead
